api/v1alpha1/meshgateway.proto

```diff
@@ -0,0 +1,119 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+syntax = "proto3";
+
+import "google/protobuf/wrappers.proto";
+import "api/v1alpha1/common.proto";
+import "gogoproto/gogo.proto";
+import "google/api/field_behavior.proto";
+import "k8s.io/api/core/v1/generated.proto";
+
+// $schema: istio-operator.api.v1alpha1.MeshGatewaySpec
+// $title: Istio Mesh Gateway Spec
+// $description: Istio Mesh Gateway descriptor
+
+package istio_operator.v2.api.v1alpha1;
+
+option go_package = "github.com/banzaicloud/istio-operator/v2/api/v1alpha1";
+
+// MeshGateway defines an Istio ingress or egress gateway
+//
+// <!-- crd generation tags
+// +cue-gen:MeshGateway:groupName:servicemesh.cisco.com
+// +cue-gen:MeshGateway:version:v1alpha1
+// +cue-gen:MeshGateway:storageVersion
+// +cue-gen:MeshGateway:annotations:helm.sh/resource-policy=keep
+// +cue-gen:MeshGateway:subresource:status
+// +cue-gen:MeshGateway:scope:Namespaced
+// +cue-gen:MeshGateway:resource:shortNames=mgw,meshgw
+// +cue-gen:MeshGateway:printerColumn:name=Age,type=date,JSONPath=.metadata.creationTimestamp,description="CreationTimestamp is a timestamp
+// representing the server time when this object was created. It is not guaranteed to be set in happens-before order across separate operations.
+// Clients may not set this value. It is represented in RFC3339 form and is in UTC.
+// Populated by the system. Read-only. Null for lists. More info: https://git.k8s.io/community/contributors/devel/api-conventions.md#metadata"
+// +cue-gen:MeshGateway:preserveUnknownFields:false
+// -->
+//
+// <!-- go code generation tags
+// +genclient
+// +k8s:deepcopy-gen=true
+// -->
+message MeshGatewaySpec {
+    // Contains the intended version for the Mesh Gateway.
+    // +kubebuilder:validation:Pattern=^1.
+    string version = 1 [(gogoproto.jsontag) = "version", (google.api.field_behavior) = REQUIRED];
+    // General Mesh gateway configuration
+    MeshGatewayConfiguration MeshGwConfiguration = 2 [(gogoproto.nullable) = false, (gogoproto.embed) = true, (gogoproto.jsontag) = ",inline"];
+    // Service ports to open for the gateway
+    // +kubebuilder:validation:MinItems=0
+    repeated ServicePort ports = 3 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "ports"];
+    // Type of gateway, either ingress or egress
+    // +kubebuilder:validation:Enum=ingress;egress
+    GatewayType type = 4  [(gogoproto.jsontag) = "type"];
+    // Istio CR to which this gateway belongs to
+    NamespacedName istioControlPlane = 5 [(gogoproto.jsontag) = "istioControlPlane,omitempty"];
+}
```

Comment:
Are these fields generated as required fields like this? I guess not, so I'd suggest using the necessary annotation for that like this:

```suggestion
    repeated ServicePort ports = 3 [(gogoproto.nullable) = false, (gogoproto.jsontag) = "ports"], (google.api.field_behavior) = REQUIRED];
    // Type of gateway, either ingress or egress
    // +kubebuilder:validation:Enum=ingress;egress
    GatewayType type = 4  [(gogoproto.jsontag) = "type"], (google.api.field_behavior) = REQUIRED];
    // Istio CR to which this gateway belongs to
    NamespacedName istioControlPlane = 5 [(gogoproto.jsontag) = "istioControlPlane"], (google.api.field_behavior) = REQUIRED];
}
```

The `istioControlPlane` field was not a required field in the v1 version, but I think that was only for backward compatibility reasons, imho we can simply make it required as well.

Also, the same applies to all other required fields.

Reply:
What do you mean by generated as required fields? I don't think the `REQUIRED` tag is really making a difference here, but I can add them.

Reply:
It should make a difference for the CRD generation: https://github.com/banzaicloud/istio-operator/blob/v2/config/crd/bases/istio-operator-crds.gen.yaml#L1490-L1491

Reply:
and it's here: https://github.com/banzaicloud/istio-operator/pull/686/commits/20fcca190848a9c27287de1325c63d00299a9c69#diff-b3f87a119ff83e86985bcac0532f13a79e219aac37a9879c0c8b024e367082eeR2356-R2358